File: Week 8/Java/addBinary.java
/**
 * Prompt:
 * 
 * Given two binary strings a and b, return their sum as a binary string.
 * 
 * https://leetcode.com/problems/add-binary/
 */

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */

class Solution {
    public String addBinary(String a, String b) {
        return new java.math.BigInteger(a, 2) // parse from string a
                .add(new java.math.BigInteger(b, 2)) // parse from string b
                .toString(2); // convert back to binary string
    }
}

/*
 * Rather simple solution, stats are crazy either. Runtime of 5ms (faster than
 * 22.23%) and memory usage of 39.4mb (less than 17.61%)
 */

File: Week 8/JavaScript/addBinary.js
/**
 * Prompt:
 *
 * Given two binary strings a and b, return their sum as a binary string.
 *
 * https://leetcode.com/problems/add-binary/
 */

/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */

/**
 * Fairly simple solution. We use both inputs to build binary literals, calculate their sum using the BigInt function,
 * then returning the sum with a .toString() call to satisfy the expected return parameters.
 */

let addBinary = (a, b) => {
  const aBin = `0b${a}`;
  const bBin = `0b${b}`;
  const sum = BigInt(aBin) + BigInt(bBin);
  return sum.toString(2);
};

/**
 * Fantastic results, 80 ms (faster than 94.59%) and 38.9mb of memory (less than 96.37%).
 */
